test/rtl/basic.mlir

```diff
@@ -21,5 +26,10 @@ func @test1(%arg0: i3) -> i50 {
 // CHECK-NEXT:    %3 = rtl.zext %arg0 : i3, i7
 // CHECK-NEXT:    %4 = rtl.concat %c42_i12 : (i12) -> i12
 // CHECK-NEXT:    %5 = rtl.concat %4, %0, %1, %2, %3 : (i12, i12, i12, i7, i7) -> i50
+
+// CHECK-NEXT:    %6 = rtl.andr %arg0 : i3, i1
```

Comment:
Driveby nit: We should try to avoid checking the SSA value names directly. Among other things, it makes updating these types of tests really difficult. I would advise checking out: https://mlir.llvm.org/getting_started/TestingGuide/

There is also a script that can help automate the generation of CHECK lines: https://github.com/llvm/llvm-project/blob/master/mlir/utils/generate-test-checks.py

-- River

Reply:
Thanks, @River707 reading.

Reply:
> @River707
> 
> > We should try to avoid checking the SSA value names directly.
> 
> I have a test that somehow passes.
> 
> ```
>   // CHECK-NEXT:    %8 = rtl.concat %4, %0, %1, %2, %3 : (i12, i12, i12, i7, i7) -> i50
>   %result = rtl.concat %conc1, %b, %c, %d, %e : (i12, i12, i12, i7, i7) -> i50
>   // CHECK-NEXT:    return %8 : i50
>   return %result : i50
> ```

The names that you put in the test file are generally not preserved during round-trip, so they will get renumbered by the printer. Thus why you generally see %8 instead of %result in the output.

> 
> Do you recommend the usage of substitutions like this?
> 
> ```
>   // CHECK-NEXT:    %[[RESULT:.*]] = rtl.concat %4, %0, %1, %2, %3 : (i12, i12, i12, i7, i7) -> i50
>   %result = rtl.concat %conc1, %b, %c, %d, %e : (i12, i12, i12, i7, i7) -> i50
>   // CHECK-NEXT:    return %[[RESULT:.*]] : i50
>   return %result : i50
> ```
> 

Yes, the way you've defined RESULT is the type of usage that we generally like to see. There is one slight problem here though. You are actually defining RESULT twice in this block. FileCheck substitution variables are defined via `[[<NAME>:<MATCHREGEX>]]` and used via `[[<NAME>]]`. Here you've got the match regex on both the definition and your usage. See here:
https://llvm.org/docs/CommandGuide/FileCheck.html#filecheck-string-substitution-blocks

> For some reason, I can't name variables in the `CHECK-NEXT:` do you know why?
> 
> ```
>   // CHECK-NEXT:    %result = rtl.concat %4, %0, %1, %2, %3 : (i12, i12, i12, i7, i7) -> i50
>   %result = rtl.concat %conc1, %b, %c, %d, %e : (i12, i12, i12, i7, i7) -> i50
>   // CHECK-NEXT:    return %result : i50
>   return %result : i50
> ```
> 
> I am getting this error:
> 
> ```
> /home/drom/work/github/circt/circt/test/rtl/basic.mlir:27:17: error: CHECK-NEXT: expected string not found in input
>  // CHECK-NEXT: %result = rtl.concat %4, %0, %1, %2, %3 : (i12, i12, i12, i7, i7) -> i50
>                 ^
> <stdin>:14:2: note: scanning from here
>  %8 = rtl.concat %4, %0, %1, %2, %3 : (i12, i12, i12, i7, i7) -> i50
>  ^
> ```

Yes, I touched on this a bit in the response above. The SSA value name is not preserved by the compiler, so it will get renamed to whatever it sees fit(with some exceptions as dialects/operations can request a specific name prefix). This is why we use FileCheck substitution variables for the check lines instead.